python_modules/dagster/dagster/_core/definitions/assets/definition/computation.py

```diff
@@ -0,0 +1,135 @@
+from collections.abc import Iterator, Mapping, Sequence
+from typing import TYPE_CHECKING, Callable, Union
+
+from dagster_shared import check
+from dagster_shared.record import record
+from typing_extensions import TypeAlias
+
+from dagster._core.definitions.asset_checks.asset_check_result import AssetCheckResult
+from dagster._core.definitions.asset_checks.asset_check_spec import AssetCheckSpec
+from dagster._core.definitions.asset_key import AssetKey
+from dagster._core.definitions.assets.definition.asset_effect import AssetEffect, Effect
+from dagster._core.definitions.assets.definition.asset_spec import AssetSpec
+from dagster._core.definitions.assets.definition.assets_definition import AssetsDefinition
+from dagster._core.definitions.decorators.decorator_assets_definition_builder import (
+    build_and_validate_named_ins,
+)
+from dagster._core.definitions.decorators.op_decorator import DecoratedOpFunction
+from dagster._core.definitions.inference import get_config_param_type
+from dagster._core.definitions.node_definition import NodeDefinition
+from dagster._core.definitions.op_definition import OpDefinition
+from dagster._core.definitions.result import AssetResult
+from dagster.components.resolved.core_models import OpSpec
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.assets.definition.asset_dep import AssetDep
+
+ComputationFn: TypeAlias = Callable[..., Iterator[Union[AssetResult, AssetCheckResult]]]
+
+
+@record
+class Computation:
+    """Generalized class for represting a single computation that has effects on entities in the
+    asset graph.
+
+    For now, this is a simple translation layer that swiftly gets turned into an AssetsDefinition,
+    but in the future this relationship will flip, and this Computation class will take over the
+    role of AssetsDefinition in the internals of the system.
+    """
+
+    node_def: NodeDefinition
+    input_mappings: Mapping[str, AssetKey]
+    output_mappings: Mapping[str, Effect]
+
+    can_subset: bool
+
+    @staticmethod
+    def from_fn(
+        fn: ComputationFn, *, op_spec: OpSpec, effects: Sequence[Effect], can_subset: bool
+    ) -> "Computation":
+        """Create a computation from a single function. Output definitions are created from the
+        provided effects. Input definitions are inferred from the function signature and provided
+        dependencies. The underlying node will be an OpDefinition that wraps the provided function.
+
+        Args:
+            fn: The function to create a computation from.
+            op_spec: The spec for the op that will be created.
+            effects: The effects that the function will have when executed.
+            can_subset: Whether the computation can be subsetted.
```

Comment:
curious if you have context on why we didn't end up putting this in the op spec historically

Reply:
philosophically I get that it's sort of both a property of the op and the assets, but I can't really concoct a scenario where you'd have an otherwise subsettable op which would become no longer subsettable when specific assets are bound to it - though I guess "subset" is just sort of a vacuous notion in an op-without-specs world

Reply:
> though I guess "subset" is just sort of a vacuous notion in an op-without-specs world

Yeah that's right -- essentially the thing here is that the "subset" operation is highly specific to the presence of asset graph entities being affected by the underlying operation.

The mechanism for doing a subsetted execution is to pass in a set of asset / check keys through the context and have the underlying compute do the minimum amount of work necessary to just affect those.